myst_nb/read.py

```diff
@@ -0,0 +1,371 @@
+
+    :returns: the notebook reader, and the (potentially modified) MdParserConfig,
+        or None if the input cannot be read as a notebook.
+    """
+    # the import is here so this module can be loaded without sphinx
+    from sphinx.util import import_object
+
+    # get all possible readers
+    readers = nb_config.custom_formats.copy()
+    # add the default reader
+    readers.setdefault(".ipynb", (standard_nb_read, {}, False))
+
+    # we check suffixes ordered by longest first, to ensure we get the "closest" match
+    iterator = sorted(readers.items(), key=lambda x: len(x[0]), reverse=True)
+    for suffix, (reader, reader_kwargs, commonmark_only) in iterator:
+        if path.endswith(suffix):
+            if isinstance(reader, str):
+                # attempt to load the reader as an object path
+                reader = import_object(reader)
+            if commonmark_only:
+                # Markdown cells should be read as Markdown only
+                md_config = attr.evolve(md_config, commonmark_only=True)
+            return NbReader(partial(reader, **(reader_kwargs or {})), md_config)
+
+    # a Markdown file is a special case, since we only treat it as a notebook,
+    # if it starts with certain "top-matter"
+    if content is not None and is_myst_markdown_notebook(content):
+        return NbReader(
+            partial(
+                read_myst_markdown_notebook,
+                config=md_config,
+                add_source_map=True,
+                path=path,
+            ),
+            md_config,
+        )
+
+    # if we get here, we did not find a reader
+    return None
+
+
+def is_myst_markdown_notebook(text: Union[str, Iterator[str]]) -> bool:
+    """Check if the input is a MyST Markdown notebook.
+
+    This is identified by the presence of a top-matter section, containing::
+
+        ---
+        jupytext:
+            text_representation:
+                format_name: myst
+        ---
+
+    :param text: The input text.
+    :returns: True if the input is a markdown notebook.
+    """
+    if isinstance(text, str):
+        if not text.startswith("---"):  # skip creating the line list in memory
+            return False
+        text = (line for line in text.splitlines())
+    try:
+        if not next(text).startswith("---"):
+            return False
+    except StopIteration:
+        return False
+    top_matter = []
+    for line in text:
+        if line.startswith("---") or line.startswith("..."):
+            break
+        top_matter.append(line.rstrip() + "\n")
+    try:
+        metadata = yaml.safe_load("".join(top_matter))
+        assert isinstance(metadata, dict)
+    except Exception:
+        return False
+    if (
+        metadata.get("jupytext", {})
+        .get("text_representation", {})
+        .get("format_name", None)
+        != "myst"
+    ):
+        return False
+
+    return True
+
+    # TODO move this to reader, since not strictly part of function objective
```

Comment:
Just flagging in case you missed this - unclear if it's TODO in this PR or TODO in the future

Reply:
I think this does not need to be strictly addressed in this PR